File: typescript/kata1/src/ShowScheduler/ShowScheduler.test.ts
import { Show } from "../Show/Show.js";
import { ShowScheduler } from "./ShowScheduler.js";

const blindWonderShow = new Show("The Blind Wonder", "tightrope");
blindWonderShow.ropeLength = 20;
blindWonderShow.isBlindfolded = true;

const funnyJoeShow = new Show("Funny Joe", "clown");
funnyJoeShow.jokes = ["joke about Chuck Norris 1", "joke about Chuck Norris 2"];

const dingoShow = new Show("Dingo", "clown");
dingoShow.jokes = [
  "JS developers joke",
  "backend developers joke",
  "senior developers joke",
];

const daredevilWalkerShow = new Show("The Daredevil Walker", "tightrope");
daredevilWalkerShow.ropeLength = 20;
daredevilWalkerShow.isBlindfolded = false;

const flyingStarsShow = new Show("The Flying Stars", "trapeze");
flyingStarsShow.performers = 2;

const threeAmigosShow = new Show("The Three Amigos", "trapeze");
threeAmigosShow.performers = 3;

describe("ShowScheduler", () => {
  it("calculates total duration for a clown show", () => {
    const shows = [dingoShow];
    const scheduler = new ShowScheduler(shows);

    const totalDuration = scheduler.calculateTotalDuration();

    expect(totalDuration).toBe(15);
  });

  it("calculates total duration for a tightrope show without blindfold", () => {
    const shows = [daredevilWalkerShow];
    const scheduler = new ShowScheduler(shows);

    const totalDuration = scheduler.calculateTotalDuration();

    expect(totalDuration).toBe(10);
  });

  it("calculates total duration for a tightrope show with blindfold", () => {
    const shows = [blindWonderShow];
    const scheduler = new ShowScheduler(shows);

    const totalDuration = scheduler.calculateTotalDuration();

    expect(totalDuration).toBe(15);
  });

  it("calculates total duration for a trapeze show", () => {
    const shows = [threeAmigosShow];
    const scheduler = new ShowScheduler(shows);

    const totalDuration = scheduler.calculateTotalDuration();

    expect(totalDuration).toBe(18);
  });

  it("calculates total duration for multiple shows", () => {
    const shows = [funnyJoeShow, blindWonderShow, flyingStarsShow];

    const scheduler = new ShowScheduler(shows);

    const totalDuration = scheduler.calculateTotalDuration();

    expect(totalDuration).toBe(37);
  });

  it("calculates duration for an existent show by its name", () => {
    const shows = [funnyJoeShow, blindWonderShow, flyingStarsShow];
    const scheduler = new ShowScheduler(shows);

    const duration = scheduler.getShowDurationByName("Funny Joe");

    expect(duration).toBe(10);
  });

  it("throws when calculating duration for a non existent show by its name", () => {
    const shows = [funnyJoeShow, blindWonderShow, flyingStarsShow];
    const scheduler = new ShowScheduler(shows);

    expect(() => {
      scheduler.getShowDurationByName("The Great FireEater");
    }).toThrow();
  });
});

File: java/kata1/src/test/java/com/therefactorproject/showscheduler/ShowSchedulerTest.java
package com.therefactorproject.showscheduler;

import com.therefactorproject.show.Show;
import org.junit.jupiter.api.Test;

import java.util.List;

import static com.therefactorproject.show.Show.show;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

class ShowSchedulerTest {

    private final Show blindWonderShow = Show.createTightropeShow("The Blind Wonder", 20.0, true);
    private final Show funnyJoeShow = Show.createClownShow("Funny Joe", List.of("joke about Chuck Norris 1", "joke about Chuck Norris 2"));
    private final Show dingoShow = Show.createClownShow("Dingo", List.of("JS developers joke", "backend developers joke", "senior developers joke"));
    private final Show daredevilWalkerShow = Show.createTightropeShow("The Daredevil Walker", 20.0, false);
    private final Show flyingStarsShow = Show.createTrapezeShow("The Flying Stars", 2);
    private final Show threeAmigosShow = Show.createTrapezeShow("The Three Amigos", 3);

    @Test
    void calculatesTotalDurationForClownShow() {
        var scheduler = new ShowScheduler(List.of(dingoShow));

        var totalDuration = scheduler.calculateTotalDuration();

        assertEquals(15, totalDuration);
    }

    @Test
    void calculatesTotalDurationForTightropeShowWithoutBlindfold() {
        var scheduler = new ShowScheduler(List.of(daredevilWalkerShow));

        var totalDuration = scheduler.calculateTotalDuration();

        assertEquals(10, totalDuration);
    }

    @Test
    void calculatesTotalDurationForTightropeShowWithBlindfold() {
        var scheduler = new ShowScheduler(List.of(blindWonderShow));

        var totalDuration = scheduler.calculateTotalDuration();

        assertEquals(15, totalDuration);
    }

    @Test
    void calculatesTotalDurationForTrapezeShow() {
        var scheduler = new ShowScheduler(List.of(threeAmigosShow));

        var totalDuration = scheduler.calculateTotalDuration();

        assertEquals(18, totalDuration);
    }

    @Test
    void calculatesTotalDurationForMultipleShows() {
        var scheduler = new ShowScheduler(List.of(funnyJoeShow, blindWonderShow, flyingStarsShow));

        var totalDuration = scheduler.calculateTotalDuration();

        assertEquals(37, totalDuration);
    }

    @Test
    void calculatesDurationForExistentShowByName() {
        var scheduler = new ShowScheduler(List.of(funnyJoeShow, blindWonderShow, flyingStarsShow));

        var duration = scheduler.getShowDurationByName("Funny Joe");

        assertEquals(10, duration);
    }

    @Test
    void throwsWhenCalculatingDurationForNonExistentShowByName() {
        var scheduler = new ShowScheduler(List.of(funnyJoeShow, blindWonderShow, flyingStarsShow));

        assertThrows(IllegalArgumentException.class, () -> scheduler.getShowDurationByName("The Great FireEater"));
    }

  @Test
  void calculateZeroDurationFoUnknownType() {
    var scheduler = new ShowScheduler(List.of(show("any", "unknown")));

    assertEquals(0, scheduler.calculateTotalDuration());
  }

}
